File: src/main/java/br/com/pnp/compiler/semantic/Analyser.java
package br.com.pnp.compiler.semantic;

import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.tree.ParseTreeWalker;
import br.com.pnp.compiler.exception.CompilationException;
import br.com.pnp.compiler.exception.SemanticException;
import br.com.pnp.compiler.model.construct.Block;
import br.com.pnp.compiler.model.construct.Construct;
import br.com.pnp.compiler.model.construct.statement.StatementBlock;
import br.com.pnp.compiler.model.expression.Expression;
import br.com.pnp.compiler.model.construct.Variable;
import br.com.pnp.compiler.model.instruction.Instruction;
import br.com.pnp.compiler.syntax.grammar.antlr.PnpLexer;
import br.com.pnp.compiler.syntax.grammar.antlr.PnpParser;

import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Stack;

public class Analyser {
    private Stack<Expression> executionStack = new Stack<Expression>();
    private SymbolTable mainTable = new SymbolTable();
    private SymbolTable currentTable = mainTable;
    private Stack<Block> blocks = new Stack<>();
    private Block currentBlock = null;

    public void analyse(String sourceFile) throws CompilationException {
        try {
            PnpLexer lexical = new PnpLexer(CharStreams.fromFileName(sourceFile));
            CommonTokenStream tokens = new CommonTokenStream(lexical);
            PnpParser syntax = new PnpParser(tokens);
            PnpParser.FileContext tree = syntax.file();

            PnpContext rules = new PnpContext(this);
            ParseTreeWalker verifier = new ParseTreeWalker();
            verifier.walk(rules, tree);
        }
        catch (RecognitionException | SemanticException ex) {
            throw new CompilationException(ex.getMessage());
        } catch (IOException ex) {
            System.out.println("FATAL: não achei o fonte " + sourceFile);
            System.exit(1);
        }
    }

    public SymbolTable getMainSymbolTable() {
        return mainTable;
    }

    public boolean tryPush(Expression item) {
        try {
            executionStack.push(item);
            return true;
        }
        catch (Exception e) {
            return false;
        }
    }

    public Expression tryPop() {
        if (!executionStack.empty()) {
            return executionStack.pop();
        }
        return null;
    }

    public Collection<Construct> tryGetValues() {
        return currentTable.getValues();
    }

    public Construct tryGetConstruct(String key) {
        return currentTable.tryGetValue(key);
    }

    public void tryPutConstruct(String key, Construct value) {
        currentTable.tryPutValue(key, value);
    }

    public boolean exists(String key) {
        return currentTable.exists(key);
    }

    public boolean existsInThisScope(String key) {
        return currentTable.existsInThisScope(key);
    }

    public List<Expression> getList() {
        List<Expression> test = (List<Expression>) executionStack.clone();
        return test;
    }

    public Block getCurrentBlock() {
        return currentBlock;
    }

    public void newAssignment(Variable variable, Expression expression) {
        if (currentBlock != null) {
            currentBlock.addAssignmentCommand(variable, expression);
        }
    }

    public void newInstruction(Instruction statement) {
        if (currentBlock != null) {
            currentBlock.addInstruction(statement);
        }
    }

    public Instruction lastInstruction() {
        if (currentBlock != null) {
            return currentBlock.getLastInstruction();
        }
        return null;
    }

    public void newScope(Block newBlock) {
        blocks.push(currentBlock);
        currentBlock = newBlock;
        currentTable = currentTable.startNewScope();
    }

    public void newScope() {
        blocks.push(currentBlock);
        currentBlock = new StatementBlock();
        currentTable = currentTable.startNewScope();
    }

    public void endScope() {
        if (currentBlock != null) {
            for(Variable var : currentTable.getVariables()) {
                currentBlock.addDeclarationCommand(var);
            }
        }
        currentTable = currentTable.lastScope();
        if (blocks.empty()) {
            currentBlock = null;
        }
        else {
            currentBlock = blocks.pop();
        }
    }
}

File: src/main/kotlin/br/com/pnp/semantic/Analyser.kt
package br.com.pnp.semantic

import br.com.pnp.exception.CompilationException
import br.com.pnp.exception.SemanticException
import br.com.pnp.grammar.antlr.PnpLexer
import br.com.pnp.grammar.antlr.PnpParser
import br.com.pnp.model.construct.Block
import br.com.pnp.model.construct.Construct
import br.com.pnp.model.construct.Variable
import br.com.pnp.model.construct.statement.StatementBlock
import br.com.pnp.model.expression.Expression
import br.com.pnp.model.instruction.Instruction
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.tree.ParseTreeWalker
import java.io.File
import java.io.IOException
import java.util.EmptyStackException
import java.util.Stack

class Analyser {
    val mainSymbolTable = SymbolTable()
    private var currentSymbolTable = mainSymbolTable
    private val executionStack = Stack<Expression>()
    private val blocks = Stack<Block>()

    val currentBlock: Block?
        get() {
            return try {
                blocks.peek()
            } catch (e: EmptyStackException) {
                null
            }
        }
    val lastInstruction: Instruction?
        get() = currentBlock?.lastInstruction

    fun analyse(file: File) {
        try {
            analyse(CharStreams.fromFileName(file.absolutePath))
        } catch (ex: IOException) {
            println(ex.message)
        }
    }

    fun analyse(sourceCode: String) {
        analyse(CharStreams.fromString(sourceCode))
    }

    private fun reset() {
        mainSymbolTable.resetTable()
        currentSymbolTable = mainSymbolTable
        executionStack.clear()
        blocks.clear()
    }

    private fun analyse(source: CharStream) {
        try {
            val lexical = PnpLexer(source)
            val tokens = CommonTokenStream(lexical)
            val syntax = PnpParser(tokens)

            val tree = syntax.file()
            val rules = PnpContext(this)

            reset()
            ParseTreeWalker().walk(rules, tree)
        } catch (ex: RecognitionException) {
            throw CompilationException(ex.message)
        } catch (ex: SemanticException) {
            throw CompilationException(ex.message)
        }
    }

    fun tryPush(item: Expression): Boolean {
        return try {
            executionStack.push(item)
            true
        } catch (e: Exception) {
            false
        }
    }

    fun tryPop(): Expression? {
        return try {
            executionStack.pop()
        } catch (ex: EmptyStackException) {
            null
        }
    }

    fun tryPeek(): Expression? {
        return try {
            executionStack.peek()
        } catch (ex: EmptyStackException) {
            null
        }
    }

    fun tryGet(key: String): Construct? {
        return currentSymbolTable.tryGetValue(key)
    }

    fun tryPut(value: Construct): Boolean {
        return currentSymbolTable.tryPutValue(value.name, value)
    }

    fun exists(key: String): Boolean {
        return currentSymbolTable.exists(key)
    }

    fun existsInScope(key: String): Boolean {
        return currentSymbolTable.existsInThisScope(key)
    }

    fun newAssignment(variable: Variable, expression: Expression): Boolean {
        return currentBlock?.run {
            if (exists(variable.name)) {
                addAssignmentInstruction(variable, expression)
                true
            } else {
                false
            }
        } ?: false
    }

    fun newInstruction(instruction: Instruction): Boolean {
        return currentBlock?.apply {
            addInstruction(instruction)
        } != null
    }

    fun newScope(newScope: Block? = null) {
        blocks.push(newScope ?: StatementBlock())
        currentSymbolTable = currentSymbolTable.startNewScope()
    }

    fun endScope() {
        for (variable in currentSymbolTable.variables) {
            currentBlock?.addDeclarationInstruction(variable)
        }

        currentSymbolTable.lastScope()?.let {
            currentSymbolTable = it
        }

        if (!blocks.empty()) {
            blocks.pop()
        }
    }
}
